src/CMakeLists.txt

```diff
@@ -1,549 +1,534 @@
+endif(CCACHE_FOUND)
+
+set(RESOURCE_FILES
+    breeze-qownnotes.qrc
+    breeze-dark-qownnotes.qrc
+    qownnotes.qrc
+    demonotes.qrc
+    images.qrc
+    texts.qrc
+    configurations.qrc
+    libraries/qmarkdowntextedit/media.qrc
+    libraries/qdarkstyle/style.qrc
+    libraries/qkeysequencewidget/qkeysequencewidget/qkeysequencewidget.qrc
+)
+
+set(SOURCE_FILES
+    dialogs/welcomedialog.cpp
+    dialogs/welcomedialog.h
+    dialogs/welcomedialog.ui
+    dialogs/issueassistantdialog.cpp
+    dialogs/issueassistantdialog.h
+    dialogs/issueassistantdialog.ui
+    dialogs/aboutdialog.cpp
+    dialogs/aboutdialog.h
+    dialogs/aboutdialog.ui
+    dialogs/linkdialog.cpp
+    dialogs/linkdialog.h
+    dialogs/linkdialog.ui
+    dialogs/notediffdialog.cpp
+    dialogs/notediffdialog.h
+    dialogs/notediffdialog.ui
+    dialogs/settingsdialog.cpp
+    dialogs/settingsdialog.h
+    dialogs/settingsdialog.ui
+    dialogs/tododialog.cpp
+    dialogs/tododialog.h
+    dialogs/tododialog.ui
+    dialogs/trashdialog.cpp
+    dialogs/trashdialog.h
+    dialogs/trashdialog.ui
+    dialogs/localtrashdialog.cpp
+    dialogs/localtrashdialog.h
+    dialogs/localtrashdialog.ui
+    dialogs/updatedialog.cpp
+    dialogs/updatedialog.h
+    dialogs/updatedialog.ui
+    dialogs/versiondialog.cpp
+    dialogs/versiondialog.h
+    dialogs/versiondialog.ui
+    dialogs/passworddialog.cpp
+    dialogs/passworddialog.h
+    dialogs/passworddialog.ui
+    dialogs/masterdialog.cpp
+    dialogs/masterdialog.h
+    dialogs/tagadddialog.cpp
+    dialogs/tagadddialog.h
+    dialogs/tagadddialog.ui
+    dialogs/sharedialog.cpp
+    dialogs/sharedialog.h
+    dialogs/sharedialog.ui
+    dialogs/evernoteimportdialog.cpp
+    dialogs/evernoteimportdialog.h
+    dialogs/evernoteimportdialog.ui
+    dialogs/serverbookmarksimportdialog.cpp
+    dialogs/serverbookmarksimportdialog.h
+    dialogs/serverbookmarksimportdialog.ui
+    dialogs/orphanedimagesdialog.cpp
+    dialogs/orphanedimagesdialog.h
+    dialogs/orphanedimagesdialog.ui
+    dialogs/orphanedattachmentsdialog.cpp
+    dialogs/orphanedattachmentsdialog.h
+    dialogs/orphanedattachmentsdialog.ui
+    dialogs/scriptrepositorydialog.cpp
+    dialogs/scriptrepositorydialog.h
+    dialogs/scriptrepositorydialog.ui
+    dialogs/dictionarymanagerdialog.cpp
+    dialogs/dictionarymanagerdialog.h
+    dialogs/dictionarymanagerdialog.ui
+    dialogs/actiondialog.cpp
+    dialogs/actiondialog.h
+    dialogs/actiondialog.ui
+    dialogs/tabledialog.cpp
+    dialogs/tabledialog.h
+    dialogs/tabledialog.ui
+    dialogs/notedialog.cpp
+    dialogs/notedialog.h
+    dialogs/notedialog.ui
+    dialogs/websockettokendialog.cpp
+    dialogs/websockettokendialog.h
+    dialogs/websockettokendialog.ui
+    dialogs/imagedialog.cpp
+    dialogs/imagedialog.h
+    dialogs/imagedialog.ui
+    dialogs/attachmentdialog.cpp
+    dialogs/attachmentdialog.h
+    dialogs/attachmentdialog.ui

```

Comment:
no this is too long. If we are going to revise our cmake, lets use cmake `include` and file globbing to make our lives easier. We should do this for all main directories i.e., `dialogs`, `entities`, `helpers`, `api`, `widgets` and so on.

Reply:
File globbing is very bad thing from my point of view. You can use it only in initial phase of converting project from qmake or Autotools. I've saw using glob only in very bad written tutorials.  

But you right it is good idea to use `src/CMakeLists.txt` only for setup project and filling actual sources from subfolders.
I'm myself prefer using `add_subdirectory` for this instead of `include`. Compare:
```cmake
add_subdirectory(dialogs)
# vs
include(dialogs/some_cmake_name.cmake)
```

`target_sources` if it called after `add_executable` even from inner `CMakeLists.txt` could see that target.
So `dialogs/CMakeList.txt` will look like:
```cmake
target_sources(QOwnNotes PRIVATE
    dialogs/imagedialog.ui
    dialogs/attachmentdialog.cpp
    dialogs/attachmentdialog.h
    dialogs/attachmentdialog.ui
)
```
Here you can see "problem" with `target_sources`. Before CMake 3.13 `target_sources` does not support relative to `CMakeLists.txt` sources file names. Before 3.13 it think that file name should be relative to target they pointing.
So instead of require user to have CMake 3.13 there two methods to overcome this.
One is using `${CMAKE_CURRENT_LIST_DIR}`:
```cmake
set(SRC ${CMAKE_CURRENT_LIST_DIR})
target_sources(QOwnNotes PRIVATE
    ${SRC}/imagedialog.ui
    ${SRC}/attachmentdialog.cpp
    ${SRC}/attachmentdialog.h
    ${SRC}/attachmentdialog.ui
)
```
This allows you to move lines between folders without modifying folder name.
But I do not like how it looks like.
There another method. Is to use macro `target_sources_local` which will works like `target_sources` from CMake 3.13.
It will be placed into `cmake` folder and included either in main or root `CMakeLists.txt`.
Here my `TargetSourcesLocal.cmake` source:
```cmake
function(target_sources_local target)
  if(POLICY CMP0076)
    # New behavior is available, so just forward to it by ensuring
    # that we have the policy set to request the new behavior, but
    # don't change the policy setting for the calling scope
    cmake_policy(PUSH)
    cmake_policy(SET CMP0076 NEW)
    target_sources(${target} ${ARGN})
    cmake_policy(POP)
    return()
  endif()

  # Must be using CMake 3.12 or earlier, so simulate the new behavior
  unset(_srcList)
  get_target_property(_targetSourceDir ${target} SOURCE_DIR)

  foreach(src ${ARGN})
    if(NOT src STREQUAL "PRIVATE" AND
       NOT src STREQUAL "PUBLIC" AND
       NOT src STREQUAL "INTERFACE" AND
       NOT IS_ABSOLUTE "${src}")
      # Relative path to source, prepend relative to where target was defined
      file(RELATIVE_PATH src "${_targetSourceDir}" "${CMAKE_CURRENT_LIST_DIR}/${src}")
    endif()
    list(APPEND _srcList ${src})
  endforeach()
  target_sources(${target} ${_srcList})
endfunction()
```
In main or root `CMakeLists.txt` we will add it:
```cmake
include(TargetSourcesLocal)
```
After that `src/dialogs/CMakeLists.txt` will looks like:
```cmake
target_sources_local(QOwnNotes PRIVATE
    imagedialog.ui
    attachmentdialog.cpp
    attachmentdialog.h
    attachmentdialog.ui
)
```
Which is much cleaner

Reply:
> File globbing is very bad thing from my point of view. 

Really? I am curious as to why its bad :thinking: Are there any negative points?

> I'm myself prefer using `add_subdirectory` for this instead of `include`. Compare:

I use `add_subdirectory` when i want to make the subdirectory a static library, otherwise I stick to `include()`. Is there a way I can use `add_subdirectory()` without having to link it to some target?


> One is using `${CMAKE_CURRENT_LIST_DIR}`:

Yes, I learned this the hard way :laughing: 

> Which is much cleaner

Agreed. However it makes things more difficult to understand for someone who is not very good at cmake.

Btw, I just create `.cmake` files for all the subdirectories. Basically its just like this:

```cmake
file(GLOB dialogs_sources ${CMAKE_CURRENT_LIST_DIR}/*.cpp)
file(GLOB dialogs_headers ${CMAKE_CURRENT_LIST_DIR}/*.h)
file(GLOB dialogs_ui ${CMAKE_CURRENT_LIST_DIR}/*.h)

set(Dialogs ${dialogs_sources} ${dialogs_headers} $[dialogs_ui})

# same for other dirs, like entities/ utils/ so on.

## In main CMakeLists.txt file

set(SOURCES ${Dialogs} ${Entities} ${Utils} ...)
```

What do you think of this approach? It seems very simple and straightforward to me.

Reply:
> Really? I am curious as to why its bad  Are there any negative points?

I don't remember link to detailed description of problems. Of course most of these problems related to good code style. 
One problem related to performance is that speed of `file(GLOB` is not fast as you may think. It is far more slower that system find utility.
Another problem is then you add or remove files CMake will **NOT** re-run (since `file(GLOB` is slow it runs only once).
This particularly problematic when you checkout between git branches where file list could be changed but CMake will **NOT** re-run. In CMake 3.12 `CONFIGURE_DEPENDS` option tried to solve this but it will slow down CMake running almost 10 times and **it not supported** by some code generators like MSVC or XCode.
Also file globbing could add files not intended for Linux platform for example. You will need to put Linux code into other folder just for one single file.
Also CMake official documentation also discourage to use `file(GLOB`:

> Note. We do not recommend using GLOB to collect a list of source files from your source tree. If no CMakeLists.txt file changes when a source is added or removed then the generated build system cannot know when to ask CMake to regenerate. The CONFIGURE_DEPENDS flag may not work reliably on all generators, or if a new generator is added in the future that cannot support it, projects using it will be stuck. Even if CONFIGURE_DEPENDS works reliably, there is still a cost to perform the check on every rebuild.

So in bottom. Since it describe you this. Next time you heard that someone try to use `file(GLOB` for source files you `should hit` him into face :)  It is easy to learn bad things and it is hard to dislearn them.

Reply:
This really good to know. I never realized it could result in so many problems.

Reply:
> I use `add_subdirectory` when i want to make the subdirectory a static library, otherwise I stick to `include()`. 

`add_subdirectory` and `include` almost same. They differs how scope of inner things.  
`include` is like `#include` in **C/C++* (remember min/max problem in MSVC).
`add_subdirectory` run *code* in isolation since it creates new scope. You use pass result variable to caller's scope using cache variables only or with `PARENT_SCOPE` in `set` command.
Some commands like `add_library`, `add_executable` make their targets visible outside their `add_subdirectory` call (they visible at any level **after** `add_subdirectory` call.

> Is there a way I can use `add_subdirectory()` without having to link it to some target?

Yes. `add_subdirectory` does not have any relation to target linking.

Reply:
> However it makes things more difficult to understand for someone who is not very good at cmake

What makes things difficult? Placing file list in its own folder?

Reply:
>Btw, I just create .cmake files for all the subdirectories. Basically its just like this:
```cmake
file(GLOB dialogs_sources ${CMAKE_CURRENT_LIST_DIR}/*.cpp)
file(GLOB dialogs_headers ${CMAKE_CURRENT_LIST_DIR}/*.h)
file(GLOB dialogs_ui ${CMAKE_CURRENT_LIST_DIR}/*.h)

set(Dialogs ${dialogs_sources} ${dialogs_headers} $[dialogs_ui})
```
`set` command without options create **local** variable. That is why you are using `include` because it does not create new scope.  
Problem can raise if you forget rename during copy-paste to another folder you will lost some sources. Since scope is same new assignment will override old one.
If you use `add_subdirectory` you need *send* variable to parent scope (without polluting cache variables only to one level up) like this:
```cmake
set(Dialogs ${dialogs_sources} ${dialogs_headers} $[dialogs_ui} PARENT_SCOPE)
```
> What do you think of this approach? It seems very simple and straightforward to me.

As I've described using `file(GLOB` is bad. So I do not like it. We use CMake at work at daily basis for many utility tools. Using `file(GLOB` and also `${PROJECT_NAME}` for target name (also common practice in awful tutorials) always create more problems than solves.
You as developer should not be lazy in some things. DRY (Don't-Repeat-Yourself) does not mean that you can you `*` for everything.

Reply:
> This really good to know. I never realized it could result in so many problems.

I do not know if book ads allowed here. But I've increased CMake understanding radically when I bought e-book **Professional CMake**. It written in simple English which good for me not English native speaker.

Reply:
None of us(the devs here) are native English speakers here. I use that book for reference sometimes, but yeah I am lazy and stop as soon as something starts working :laughing:

Reply:
You are always welcome